ethereum/contracts/IncentivizedReceiveChannel.sol

```diff
@@ -0,0 +1,173 @@
+        // returns parachainBlockHeader
+
+        // Prove that the commitment is in fact in the parachain block header
+        // require(parachainBlockHeader.commitment == commitment)
+
+        // Validate that the commitment matches the commitment contents
+        require(
+            validateCommitment(commitment, commitmentContents),
+            "invalid commitment"
+        );
+
+        // Require there is enough gas to play all messages
+        require(
+            gasleft() >=
+                commitmentContents.messages.length * MAX_GAS_PER_MESSAGE,
+            "insufficient gas for delivery of all messages"
+        );
+
+        // Require all payloads are smaller than max_payload_size
+        for (uint256 i = 0; i < commitmentContents.messages.length; i++) {
+            require(
+                commitmentContents.messages[i].payload.length <=
+                    MAX_PAYLOAD_BYTE_SIZE,
+                "message payload bytesize exceeds maximum payload size"
+            );
+        }
+        return true;
+    }
+
+    function processCommitmentContents(
+        CommitmentContents memory commitmentContents
+    ) internal {
+        for (uint256 i = 0; i < commitmentContents.messages.length; i++) {
+            // Check message nonce is correct and increment nonce for replay protection
+            Message memory message = commitmentContents.messages[i];
+            require(message.nonce == lastProcessedNonce + 1, "invalid nonce");
+
+            lastProcessedNonce = lastProcessedNonce + 1;
+
+            // Deliver the message to the destination
+            // Delivery will have fixed maximum gas allowed for the destination app.
+            address targetApplicationAddress = message.targetApplicationAddress;
+            uint256 allowedGas = MAX_GAS_PER_MESSAGE;
+            bytes memory callInput = message.payload;
+
+            bool result;
+            uint256 callInputLength = callInput.length;
+            assembly {
+                let callInputDataPointer := add(callInput, 32) // Move 32 ahead of callInput length slot to get to actual data slot (TODO: verify this works in all cases)
+
+                // Dispatch call to the receiver - it is expected to be fire and forget. If the call reverts, runs out of gas, error,
+                // etc, its the fault of the sender
+                result := call(
+                    allowedGas, // Allowed gas
+                    targetApplicationAddress, // To addr
+                    0, // No ether value being sent
+                    callInputDataPointer, // Inputs are stored at callInputDataPointer
+                    callInputLength, // Input length
+                    callInputDataPointer, // Store output over input (saves space)
+                    callInputLength // Outputs can be as long as input length
+                )
+            }
+
+            emit MessageDelivered(message.nonce, result);
+        }
+    }
+
+    function validateCommitment(
```

Comment:
The implementation here doesn't match how commitment hashes are actually produced on the Substrate side:

```
hash = keccak256(abiEncode(messages))
```

I implemented it like that for several reasons, one of them being that it should be theoretically simpler and cheaper to reproduce the hash on the Ethereum side.

Since `messages` can be made available in the calldata for the  `newParachainCommitment()` (if we specify that storage location), we can just extract the bytes from the calldata and hash it.

Reply:
yep, @denalimarsh mentioned some issues he encountered with the hashing, so still need to resolve/fix this

Reply:
agree accessing and hashing the exact bytes via calldata slots should be more efficient though yep abi.encode(messages) will be ok as an alternative if there are issues with that

Reply:
Updated the `validateCommitment` implementation in https://github.com/Snowfork/polkadot-ethereum/pull/219/commits/35478272b36387ce85ca211d150cf839f39abb57